frontend/src/WebsocketProvider.tsx

```diff
@@ -0,0 +1,83 @@
+import * as React from 'react';
+import {AppState, AppStateConsumer} from './AppStateProvider';
+import websocketHandler from './websocketHandler';
+
+type Context = {
+  send: (value: any) => void;
+};
+
+const WebsocketContext = React.createContext<Context>({
+  send: () => {},
+});
+
+export const WebsocketConsumer = WebsocketContext.Consumer;
+
+type Props = {
+  state: AppState;
+  updateState: (state: Partial<AppState>) => void;
+  children?: React.ReactNode;
+};
+const WebsocketProvider = (props: Props) => {
+  const {state, updateState, children} = props;
+  const [websocket, setWebsocket] = React.useState<WebSocket | null>(null);
+
+  // Because state/updateState are passed in and change every time something
+  // happens, we need to maintain a reference to these props to prevent stale
+  // closures which may happen if state/updateState is changed between when an
+  // event listener is registered and when it fires.
+  // https://reactjs.org/docs/hooks-faq.html#why-am-i-seeing-stale-props-or-state-inside-my-function
+  const stateRef = React.useRef(state);
+  const updateStateRef = React.useRef(updateState);
+
+  React.useEffect(() => {
+    stateRef.current = state;
+    updateStateRef.current = updateState;
+  }, [state, updateState]);
+
+  React.useEffect(() => {
```

Comment:
I'm not quite familiar with what `useEffect` does here -- would we only init the websocket once here? or would we do it every time the props update on `WebsocketProvider`?

Reply:
I believe useEffect runs on all renders by default. The second argument to useEffect (here, `[]`) is a list of dependencies that chooses when it should be run. So empty array means that this will only run once, and is equivalent to onComponentMount